### PR TITLE
[filelogreceiver]: Add ability to sort by mtime

### PR DESCRIPTION
**Pull request by BinaryFissionGames**
**Description:** <Describe what has changed.>
* Adds a new `mtime` sort type, which will sort files by their modified time
* Add a feature gate for `mtime` sort type

An optional follow-up performance improvement may be made here, to have the finder return fs.DirEntry directly to query the mtime without making an extra call to os.Stat for each file.

**Link to tracking Issue:** #27812

**Testing:**
* Added unit tests for new functionality

**Documentation:** 
* Added new `mode` parameter to filelogreceiver docs

### CODE REVIEW

File: pkg/stanza/fileconsumer/matcher/matcher.go [hunk 1]
```diff
@@ -63,7 +73,9 @@ func New(c Criteria) (*Matcher, error) {
 		}, nil
 	}
 
-	if c.OrderingCriteria.Regex == "" {
+	regexIsUsed := orderingCriteriaNeedsRegex(c.OrderingCriteria.SortBy)
+
+	if regexIsUsed && c.OrderingCriteria.Regex == "" {
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/matcher/matcher.go:78]: I think we could consolidate this bit of logic with the part below which compiles the regex. The positioning of the checks in between appear to be arbiratry.

File: pkg/stanza/fileconsumer/matcher/internal/filter/sort.go [hunk 2]
```diff
@@ -130,3 +131,50 @@ func SortTemporal(regexKey string, ascending bool, layout string, location strin
...
+	filteredValues := make([]*item, 0, len(items))
+	for _, mtimeItem := range mtimeItems {
+		filteredValues = append(filteredValues, mtimeItem.item)
+	}
+
+	return filteredValues, errs
+}
+
+func SortMtime() Option {
+	return newMtimeSortOption()
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/matcher/internal/filter/sort.go:179]: Doesn't look like we really need `newMtimeSortOption()`
```suggestion
	return mtimeSortOption{}
```

File: pkg/stanza/fileconsumer/matcher/internal/filter/sort.go [hunk 3]
```diff
@@ -130,3 +131,50 @@ func SortTemporal(regexKey string, ascending bool, layout string, location strin
 		},
 	)
 }
+
+type mtimeSortOption struct{}
+
+func newMtimeSortOption() Option {
+	return mtimeSortOption{}
+}
+
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/matcher/internal/filter/sort.go:140]:
```suggestion
```

**Review by djaglowski — COMMENTED** [review]
This looks like a nice improvement vs the previous one. Just a couple nits.

File: pkg/stanza/fileconsumer/matcher/matcher.go (hunk 1 above)
Comment (BinaryFissionGames) [reply pkg/stanza/fileconsumer/matcher/matcher.go]: Good point, I've condensed these blocks

File: pkg/stanza/fileconsumer/matcher/internal/filter/sort.go (hunk 2 above)
Comment (BinaryFissionGames) [reply pkg/stanza/fileconsumer/matcher/internal/filter/sort.go]: Yeah, we don't need it, removed newMtimeSortOption

**Review by djaglowski — APPROVED** [review]

**Pull request merged**